File: Cpp/TREE/ZigZag Tree Traversal.cpp
//https://www.geeksforgeeks.org/problems/zigzag-tree-traversal/1
class Solution{
    public:
    //Function to store the zig zag order traversal of tree in a list.
    vector <int> zigZagTraversal(Node* root)
    {
    	// Code here
    	 vector<int>ans;
       if(root==NULL)
           return ans;
       queue<Node*>q;
       q.push(root);
       bool flag=true;
       
       while(!q.empty())
       {
           int size=q.size();
           vector<int>level;
           
           for(int i=0;i<size;i++)
           {
               Node* node=q.front();
               q.pop();
               level.push_back(node->data);
               
               if(node->left!=NULL)
                   q.push(node->left);
               if(node->right!=NULL)
                   q.push(node->right);
           }
           
           if(!flag)
           {
               reverse(level.begin(),level.end());
           }
           for(auto it:level)
                    ans.push_back(it);
            
           flag=!flag;
       }
       return ans;
    }
};

File: Java/TREE/ZigZag Tree Traversal.java
//https://www.geeksforgeeks.org/problems/zigzag-tree-traversal/1
class GFG
{
    //Function to store the zig zag order traversal of tree in a list.
	ArrayList<Integer> zigZagTraversal(Node root)
	{
	    //Add your code here.
	    ArrayList<Integer> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        boolean flag = true;
        
        while (!q.isEmpty()) {
            int size = q.size();
            ArrayList<Integer> level = new ArrayList<>();
            
            for (int i = 0; i < size; i++) {
                Node node = q.poll();
                level.add(node.data);
                
                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }
            }
            
            // If the flag is false, reverse the level before adding it to the result
            if (!flag) {
                Collections.reverse(level);
            }
            
            ans.addAll(level);
            flag = !flag; // Toggle the flag for next level
        }
        
        return ans;
	}
}
